lib/os/json.c

```diff
@@ -390,6 +390,29 @@ static int arr_next(struct json_obj *json, struct json_token *value)
 	return element_token(value->type);
 }
 
+static void skip_field(struct json_obj *obj, struct json_obj_key_value *kv)
+{
+	int field_count = 1;
+
+	if (kv->value.type == JSON_TOK_OBJECT_START ||
+	    kv->value.type == JSON_TOK_ARRAY_START) {
+		while (field_count > 0 && lexer_next(&obj->lex, &kv->value)) {
+			switch (kv->value.type) {
+			case JSON_TOK_OBJECT_START:
+			case JSON_TOK_ARRAY_START:
+				field_count++;
+				break;
+			case JSON_TOK_OBJECT_END:
+			case JSON_TOK_ARRAY_END:
+				field_count--;
+				break;
+			default:
+				break;
+			}
+		}
+	}
+}
```

Comment:
Can you capture `JSON_TOK_ERROR` here and use a return value to partly detect invalid input?

```c
static int skip_field(struct json_obj *obj, struct json_obj_key_value *kv)
{
	int field_count = 1;

	if (kv->value.type == JSON_TOK_OBJECT_START ||
	    kv->value.type == JSON_TOK_ARRAY_START) {
		while (field_count > 0 && lexer_next(&obj->lex, &kv->value)) {
			switch (kv->value.type) {
			case JSON_TOK_OBJECT_START:
			case JSON_TOK_ARRAY_START:
				field_count++;
				break;
			case JSON_TOK_OBJECT_END:
			case JSON_TOK_ARRAY_END:
				field_count--;
				break;
			case JSON_TOK_ERROR:
				return -EINVAL;
			default:
				break;
			}
		}
	} else {
		return -EINVAL;
	}

	return 0;
}
```

Reply:
Good idea to check for `JSON_TOK_ERROR`!

Returning `-EINVAL` on value types other than object and array does not seem to be correct as an extra object could also be of primitive type, e.g. following code should return successfully with return code 1:

```cpp
struct test_struct_t {
    const char* str;
} ts;

const struct json_obj_descr test_descr[] = {
  JSON_OBJ_DESCR_PRIM(struct test_struct_t, str, JSON_TOK_STRING)
};

char encoded[] = "{"
    "\"other_str\":\"12\","
    "\"str\":\"abc\","
    "\"other_num\":34"
    "}\n";

int ret = json_obj_parse(encoded, sizeof(encoded) - 1, test_descr, ARRAY_SIZE(test_descr), &ts);
```

Reply:
The `default` would still continue to work? It's only in case of an errror.